dex/encrypt/encrypt.go

```diff
@@ -0,0 +1,232 @@
+type Crypter interface {
+	Encrypt(b []byte) ([]byte, error)
+	Decrypt(b []byte) ([]byte, error)
+	Serialize() ([]byte, error)
+	Close()
+}
+
+const (
+	// defaultTime is the default time parameter for argon2id key derivation.
+	defaultTime = 1
+	// defaultMem is the default memory parameter for argon2id key derivation.
+	defaultMem = 64 * 1024
+	// KeySize is the size of the encryption key.
+	KeySize = 32
+	// checkPhrase is used to ensure the integrity of the deserialized Crypter.
+	checkPhrase = "dcrdex checkphrase"
+)
+
+var (
+	// Big-endian encoder.
+	intCoder = encode.IntCoder
+	// A zeroed Key to check for un-initialized or zeroed keys.
+	zeroKey Key
+)
+
+// Key is 32 bytes.
+type Key [KeySize]byte
+
+// Salt is randomness used as part of key deriviation. This is different from
+// the salt generated during xchacha20poly1305 encryption, which is shorter.
+type Salt [KeySize]byte
+
+// newSalt is the constructor for a salt based on randomness from crypto/rand.
+func newSalt() Salt {
+	var s Salt
+	_, err := rand.Read(s[:])
+	if err != nil {
+		panic("newSalt: " + err.Error())
+	}
+	return s
+}
+
+// argonParams is a set of parameters for key derivation.
+type argonParams struct {
+	time    uint32
+	memory  uint32
+	threads uint8
+}
+
+// NewCrypter derives an encryption key from a password string.
+func NewCrypter(pw string) Crypter {
+	return newArgonPolyCrypter(pw)
+}
+
+// Deserialize deserializes the encryption key for the password.
+func Deserialize(pw string, encCrypter []byte) (Crypter, error) {
+	ver, pushes, err := encode.DecodeBlob(encCrypter)
+	if err != nil {
+		return nil, err
+	}
+	switch ver {
+	case 0:
+		return decodeArgonPoly_v0(pw, pushes)
+	default:
+		return nil, fmt.Errorf("unknown Crypter version %d", ver)
+	}
+}
+
+// argonPolyCryper is an encryption algorithm based on argon2id for key
+// derivation and xchacha20poly1305 for symmetric encryption.
+type argonPolyCrypter struct {
+	key    Key
+	salt   Salt
+	params *argonParams
+}
+
+// newArgonPolyCrypter is the constructor for an argonPolyCrypter.
+func newArgonPolyCrypter(pw string) *argonPolyCrypter {
+	pwB := []byte(pw)
+	salt := newSalt()
+	threads := uint8(runtime.NumCPU())
+	keyB := argon2.IDKey(pwB, salt[:], defaultTime, defaultMem, threads, KeySize)
+	var key Key
+	copy(key[:], keyB)
+	return &argonPolyCrypter{
+		key:  key,
+		salt: salt,
+		params: &argonParams{
+			time:    defaultTime,
+			memory:  defaultMem,
+			threads: threads,
+		},
+	}
+}
+
+// Encrypt encrypts the plaintext.
+func (c *argonPolyCrypter) Encrypt(plainText []byte) ([]byte, error) {
+	keyHash := blake256.Sum256(c.key[:])
+	boxer, err := chacha20poly1305.NewX(c.key[:])
+	if err != nil {
+		return nil, fmt.Errorf("aead error: %v", err)
+	}
+	nonceSize := boxer.NonceSize()
+	nonce := make([]byte, nonceSize)
+	_, err = rand.Read(nonce)
+	if err != nil {
+		return nil, fmt.Errorf("nonce generation error: %v", err)
+	}
+	cipherText := boxer.Seal(nil, nonce, plainText, keyHash[:])
```

Comment:
associated data is for authenticating data that you need to keep in the clear, so don't pass the hash of the symmetric key here, because there's no point to authenticating it (leave it nil, which will make it easier to authenticate something else in the future if needed).